Update app-template dependencies and add helpful scripts

brandonwkipp opened the pull request:
This PR brings `app-template` dependencies up to par with `smm.org`. In addition, this fixes the `gatsby-config` file and adds some scripting to install the `arduino-base` submodule and its dependencies. I also copied over the `depcheck` script.

To test, you should make sure you can run `yarn develop`, `yarn install:arduino-base`, and `yarn depcheck` without issue.

tnordberg reviewed (changes requested):
This is looking great. 
`yarn develop` and `yarn depcheck` are working. There's an unused `gatsby-image`dep which confirms it's working. 

I like this method of installing `arduino-base` but ran into a couple bumps. 

1. `scripts/install-arduino-base.js` conditional is inverted. Works after switching. 
2. The dependencies of `ReactSerial` don't seem to get installed. I see the `postinstall` script but don't think it ran.
3. After copying `arduino-example-page.js` into `pages/`, I confirmed this package was missing: 
```
/Users/tnordberg/Documents/SMM/Dev/tn-dev-test-app-template/src/pages/arduino-test-page.js
  11:33  error  Unable to resolve path to module 'react-scrollable-list'  import/no-unresolved
```
Beyond these issues, I think this is ready to go. Very cool :)

brandonwkipp commented on scripts/install-arduino-base.js:
```diff
@@ -0,0 +1,11 @@
+const fs = require('fs');
+
+// Setup dir structure and install submodule for arduino-base if it does not exist
+if (fs.existsSync('src/Arduino/arduino-base/ReactSerial')) {
```
Good catch!!!

brandonwkipp commented:
Per @tnordberg 's review, I updated the conditional in `scripts/install-arduino-base.js` and I simplified the `ReactSerial` dependency installation by moving that step into the `scripts/install-arduino-base.js`. That script also can intelligently install the latest `arduino-base` submodule as well as the latest `ReactSerial` dependencies.

brandonwkipp commented:
@tnordberg Could I ask for one more clean install, and simply running `yarn install:arduino-base`?

tnordberg commented:
Hey @brandonwkipp, I'm still seeing the issue. Here are the reproduction steps:

1. Pull down fresh app-template
2. `git checkout EXTECH-14/update-dependencies`
3. Run `yarn install:arduino-base`
4. Run `yarn` 
4. Run `yarn develop` -> no errors
6. Copy `ReactSerial/arduino-test-page` into `pages/` -> Errors out
```bash
/Users/tnordberg/Documents/SMM/Dev/tn-dev-test-app-template4/src/pages/arduino-test-page.js
  11:33  error  Unable to resolve path to module 'react-scrollable-list'  import/no-unresolved
```

brandonwkipp commented:
Oh, I see what's happening. My understanding of the `arduino-test-page` (jump in if I have this wrong @sanine-a) is that it's used when one is working on the `arduino-base` repository as a way of checking serial communication so that developers in the shop don't have to bother with `app-template`.

The `install:arduino-base` command in `app-template` handles the install for npm packages within `arduino-base`, but if a developer pulls out the `arduino-test-page` into `app-template`, it will have linting errors because `react-scrollable-list` is missing from the `app-template` dependencies.

So, I think there are two different ways to solve this:

* Add `react-scrollable-list` as a `devDependency` and whitelist it in the `depcheck` script, or
* Add a different example page for use with `Gatsby`

tnordberg commented:
Thanks @brandonwkipp, got it. I was forgetting about the need to use this outside `app-template`.  

Since we still need the non-gatsby version intact for @sanine-a and others agnostic to `app-template`, so I propose we split `arduino-test-page.js` into a refactored `arduino-test-page` that assumes it’s inside `ReactSerial`, and a second `gatsby-wrapper-page` that gets `cp`’d into `pages` as the last step of `install-arduino-base.js`. E.g.:
```
execSync('cp src/Arduino/arduino-base/ReactSerial/examples/gatsby/gatsby-wrapper-page.js src/pages/arduino.js’);
```

That copied minimal wrapper page just points at the non-gatsby test page:
```
import ArduinoTestPage from '../Arduino/arduino-base/ReactSerial/examples/arduino-test-page';

const ArduinoWrapperPage = () => (
  <ArduinoTestPage />
);
```

Does that make sense to you? I could quickly make those changes to `arduino-base` today. 

It would also be easy enough to append a `yarn add react-scrollable-list` to `install-arduino-base.js`, which seems like a simple solution. 

But, that doesn’t fix the issue of conflicting node projects. It's tempting to just strip out `react-scrollable-list` altogether (it's also used in SerialHOC) but we'd be cornering ourselves into never using an external package, which doesn't seem wise.

Have you already tried adding an ignorePattern to `.eslintrc.js`?

brandonwkipp commented:
I think that solution makes sense.

Regarding the `node` conflicting errors, if we don't actually install the `arduino-base` deps when its included as a submodule, there isn't anything to conflict with. So, as long as we install the `arduino-base` ReactSerial deps in the root of `app-template`, I think we're golden. I'm looking into this and will report back what I find.

brandonwkipp commented:
I unfortunately could not find an elegant solution, but I came away with some ideas for the future where publishing an `npm` package might make sense, but that is too big to do right now. So, I just went with running `yarn add react-scrollable-list`. It works, but feels pretty hacky. I'll write up some more internal issues to tackle at a later date, but I think we're good here.
